### PR TITLE
Adding RMSE With Uncertainty loss to CatBoost

### PR DESCRIPTION
**Pull request by jonasracine**
In this PR, CatBoostModel and RegressionModel were modified to support CatBoost's RMSE With Uncertainty loss function.

With this loss, CatBoost output not a single value but a [mean, variance] list. Most of the work here was concerned with wrangling output shapes, especially in the multivariate case and with output chunks > 1.

This PR depends on the original CatBoost PR and should be merged after it.

### CODE REVIEW

File: darts/models/forecasting/regression_model.py [hunk 1]
```diff
@@ -700,6 +710,60 @@ def _predict_quantiles(
...
+        chunk_len = self.output_chunk_length
+
+        # treating each component separately
+        # optimization: everything into one large block-diagonal cov matrix?
+        mu_sigma_list = [model_output[i, :, :] for i in range(shape[0])]
+
+        list_of_samples = [
+            self._rng.multivariate_normal(
+                mu_sigma[:, 0],  # mean vector
+                np.diag(mu_sigma[:, 1]),  # diagonal covariance matrix
```
[the remark this answers is not in the document]
Comment (jonasracine) [reply darts/models/forecasting/regression_model.py:757]: Done!

File: darts/models/forecasting/regression_model.py [hunk 2]
```diff
@@ -700,6 +710,60 @@ def _predict_quantiles(
...
+            return output_slice.reshape(k, chunk_len, -1)
+
+        # univariate & single-chunk output
+        if output_dim <= 2:
+            # embedding well shaped 2D output into 3D
+            model_output = np.array([model_output])
+        else:
+            model_output = model_output.transpose()
+
+        return self._normal_sampling(model_output, k)
```
[the remark this answers is not in the document]
Comment (jonasracine) [reply darts/models/forecasting/regression_model.py:741]: You're right, done.

File: darts/models/forecasting/regression_model.py [hunk 3]
```diff
@@ -700,6 +710,60 @@ def _predict_quantiles(
...
+                output_slice = model_output[:, 0]
+            else:
+                output_slice = model_output[0, :, :]
+
+            return output_slice.reshape(k, chunk_len, -1)
+
+        # univariate & single-chunk output
+        if output_dim <= 2:
+            # embedding well shaped 2D output into 3D
+            model_output = np.array([model_output])
```
[the remark this answers is not in the document]
Comment (jonasracine) [reply darts/models/forecasting/regression_model.py:737]: Ah nice! Didn't know this method existed

File: darts/models/forecasting/catboost_model.py [hunk 4]
```diff
@@ -50,8 +49,9 @@ def __init__(
             horizon `n` used in `predict()`. However, setting `output_chunk_length` equal to the forecast horizon may
             be useful if the covariates don't extend far enough into the future.
         likelihood
-            Can be set to `quantile` or 'poisson'. If set, the model will be probabilistic, allowing sampling at
-             prediction time.
+            Can be set to 'quantile', 'poisson' or 'RMSEWithUncertainty'. If set, the model will be probabilistic,
```
[the remark this answers is not in the document]
Comment (jonasracine) [reply darts/models/forecasting/catboost_model.py:52]: I agree, would be more aligned with the current names.

**Comment by hrzn** [comment]
So in view of the tutorial you posted, what sort of "knowledge uncertainty" is this version capturing? It'd be nice to document this.

File: darts/models/forecasting/regression_model.py [hunk 5]
```diff
@@ -700,6 +710,60 @@ def _predict_quantiles(
...
         return sampled
 
+    def _predict_normal(self, x: np.ndarray, num_samples: int, **kwargs) -> np.ndarray:
+        """Method intended for CatBoost's RMSEWithUncertainty loss. Returns samples
+        computed from double-valued inputs [mean, variance].
+        X is of shape (n_series * n_samples, n_regression_features)
+        """
+        k = x.shape[0]
+
+        model_output = self.model.predict(x, **kwargs)
```
[the remark this answers is not in the document]
Comment (jonasracine) [reply darts/models/forecasting/regression_model.py:720]: Done!

**Review by hrzn — APPROVED** [review]
Looks good, thanks!

**Pull request merged**